programs/psa/psa_hash.c

```diff
@@ -0,0 +1,163 @@
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#include "psa/crypto.h"
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mbedtls/build_info.h"
+#include "mbedtls/platform.h"
+
+#define HASH_ALG PSA_ALG_SHA_256
+
+#define TEST_SHA256_HASH {                                                 \
+        0x5a, 0x09, 0xe8, 0xfa, 0x9c, 0x77, 0x80, 0x7b, 0x24, 0xe9, 0x9c, 0x9c, \
+        0xf9, 0x99, 0xde, 0xbf, 0xad, 0x84, 0x41, 0xe2, 0x69, 0xeb, 0x96, 0x0e, \
+        0x20, 0x1f, 0x61, 0xfc, 0x3d, 0xe2, 0x0d, 0x5a                          \
+}
+
+const uint8_t mbedtls_test_sha256_hash[] = TEST_SHA256_HASH;
+
+const size_t mbedtls_test_sha256_hash_len =
+    sizeof(mbedtls_test_sha256_hash);
+
+#if !defined(MBEDTLS_PSA_CRYPTO_C) || !defined(PSA_WANT_ALG_SHA_256)
+int main(void)
+{
+    mbedtls_printf("MBEDTLS_PSA_CRYPTO_C and MBEDTLS_SHA256_C"
+                   "not defined.\r\n");
+    return EXIT_SUCCESS;
+}
+#else
+
+int main(void)
+{
+    uint8_t buf[] = "Hello World!";
+    psa_status_t status;
+    uint8_t hash[PSA_HASH_LENGTH(HASH_ALG)];
+    size_t hash_size;
+    psa_hash_operation_t psa_hash_operation = PSA_HASH_OPERATION_INIT;
+    psa_hash_operation_t cloned_psa_hash_operation = PSA_HASH_OPERATION_INIT;
+
+    mbedtls_printf("PSA Crypto API: SHA-256 example\n\n");
+
+    status = psa_crypto_init();
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_crypto_init failed\n");
+        psa_hash_abort(&psa_hash_operation);
+        psa_hash_abort(&cloned_psa_hash_operation);
+        return EXIT_FAILURE;
+    }
+
+    /* Compute hash using multi-part operation */
+
+    status = psa_hash_setup(&psa_hash_operation, HASH_ALG);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_setup failed\n");
+        psa_hash_abort(&psa_hash_operation);
+        psa_hash_abort(&cloned_psa_hash_operation);
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_update(&psa_hash_operation, buf, sizeof(buf));
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_update failed\n");
+        psa_hash_abort(&psa_hash_operation);
+        psa_hash_abort(&cloned_psa_hash_operation);
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_clone(&psa_hash_operation, &cloned_psa_hash_operation);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("PSA hash clone failed");
+        psa_hash_abort(&psa_hash_operation);
+        psa_hash_abort(&cloned_psa_hash_operation);
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_finish(&psa_hash_operation, hash, sizeof(hash), &hash_size);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_finish failed\n");
+        psa_hash_abort(&psa_hash_operation);
+        psa_hash_abort(&cloned_psa_hash_operation);
+        return EXIT_FAILURE;
+    }
```

Comment:
Here would be a good point to check that `hash_size` and `hash` have the expected content.